url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/PartsRobotAgent.java

```diff
@@ -0,0 +1,531 @@
+			status = PartsRobotStatus.PICKING_UP;
+			List<Nest> readyNests = new ArrayList<Nest>(GoodParts.keySet());
+			// synchronized (GoodParts) {
+			for (Nest nest : readyNests) {
+				// A nest contains a single part type
+				Part part = GoodParts.get(nest).get(0);
+				synchronized (MyKits) {
+					// print("Size of MyKits: " +
+					// MyKits.size());
+					for (MyKit mk : MyKits) {
+						// Checking if the good part is needed by
+						// either kit
+						// print("Kit needs: " +
+						// mk.kit.partsExpected.getConfig().toString());
+						if (NumTotalPartsNeeded(part) > NumPartsInHand(part)) {
+							print("Found a part I need of type " + part.type.getName() + " for kit "
+									+ MyKits.indexOf(mk) + " " + mk.kit.PartsStillNeeded());
+							synchronized (Arms) {
+								for (Arm arm : Arms) {
+									if (arm.AS == ArmStatus.EMPTY) {
+										// Find the empty arm
+										PickUpPart(arm, part, nest);
+										return true;
+									}
+								}
+							}
+
+						} // Why is this so awful
+					}
+				}
+			}
+			// }
+		} else {
+			status = PartsRobotStatus.PLACING;
+		}
+
+		if (System.currentTimeMillis() - time.getTime() > 5000) {
+			// Last rule is to place parts if the parts robot has been idle
+			// too long
+			time.setTime(System.currentTimeMillis());
+			status = PartsRobotStatus.PLACING;
+			return true;
+		}
+
+		timer.schedule(new TimerTask() {
+			// hack to force the partsrobot to attempt to place parts
+			// sleep. Fires after 5.001 seconds.
+			@Override
+			public void run() {
+				stateChanged();
+			}
+		}, 5001);
+
+		/*
+		 * Tried all rules and found no actions to fire. Return false to the main loop of abstract base class Agent and
+		 * wait.
+		 */
+		return false;
+	}
+
+	/*
+	 * Actions
+	 */
+
+	private void PickUpPart(Arm arm, Part part, Nest nest) {
+		time.setTime(System.currentTimeMillis());
+		print("Picking up part" + getPartTypesInArms());
+		// synchronized (Arms) {
+
+		GoodParts.remove(nest);
+		arm.AS = ArmStatus.FULL;
+		arm.part = part;
+		// Tells the graphics to pickup the part
+		if (partsRobotGraphics != null) {
+			partsRobotGraphics.pickUpPart(part.partGraphics, Arms.indexOf(arm));
+			try {
+				// print("Blocking");
+				animation.acquire();
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
+			// print("Got permit");
+		}
+
+		// Only takes 1 part from a nest at a time
+		nest.msgTakingPart(part);
+		nest.msgDoneTakingParts();
+
+		stateChanged();
+		// }
+	}
+
+	private void PlacePart(Arm arm) {
+
+		print("Placing part");
+		synchronized (MyKits) {
+			for (MyKit mk : MyKits) {
+				if (mk.kit.needPart(arm.part) > 0) {
+					if (Math.random() <= dropChance) {
+						DropPart(arm);
+					}
+					if (partsRobotGraphics != null) {
+						partsRobotGraphics.givePartToKit(arm.part.partGraphics, mk.kit.kitGraphics, Arms.indexOf(arm));
+						try {
+							// print("Blocking");
+							animation.acquire();
+						} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)